Java/core/src/main/java/com/amazon/randomcutforest/preprocessor/ImputePreprocessor.java

```diff
@@ -168,7 +142,9 @@ void updateForest(boolean changeForest, double[] input, long timestamp, RandomCu
         updateShingle(input, scaledInput);
         updateTimestamps(timestamp);
         if (isFullyImputed) {
-            numberOfImputed = numberOfImputed + 1;
+            numberOfImputed = Math.min(numberOfImputed + 1, shingleSize);
+        } else if (numberOfImputed > 0) {
+            numberOfImputed = numberOfImputed - 1;
```

Comment:
at this point are we guaranteed to have seen a new value that is not imputed?

Reply:
yes.